websocket: remove consts in websocket_request structure

joelguittet opened the pull request:
Remove const definitions in the websocket_request structure so that user application is not constrained by the definition of the fields.

@jukkar this is following discussion on Discord. Do you think this should be backported to v3.7 branch ? This should be better according to me.

joelguittet committed:
net: websocket: remove consts in websocket_request structure

Remove const definitions in the websocket_request structure so that user
application is not constrained by the definition of the fields.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

jukkar commented on include/zephyr/net/websocket.h:
```diff
@@ -69,10 +69,10 @@ typedef int (*websocket_connect_cb_t)(int ws_sock, struct http_request *req,
  */
```
Looks good, minor nit about the commit subject, please add net there so `net: websocket: ...`

jukkar commented:
> Do you think this should be backported to v3.7 branch ? This should be better according to me.

Sure, we can backport that 3.7. You need to create an issue in this case and then reference to it in the commit message.

joelguittet commented:
> Sure, we can backport that 3.7. You need to create an issue in this case and then reference to it in the commit message.

Can you point me an example ? Or documentation maybe ? I'm not used to do this, this will be the first time

jukkar commented:
> Can you point me an example ? Or documentation maybe ? I'm not used to do this, this will be the first time

For example this one https://github.com/zephyrproject-rtos/zephyr/issues/78146 which I created yesterday for backporting purposes.

joelguittet committed:
net: http: remove consts in http_request structure

Remove const definitions in the http_request structure so that user
application is not constrained by the definition of the fields.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

rlubos commented on include/zephyr/net/http/client.h:
```diff
@@ -254,27 +254,27 @@ struct http_request {
 
 	/** The URL for this request, for example: /index.html */
-	const char *url;
+	char *url;
```
I'm a bit lost here, how exactly the application is constrained by those fields being "const"? I have an impression it's the other way around, if we don't guarantee the library won't touch those fields, we couldn't assign URL constants as we do in the [HTTP client sample](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/sockets/http_client/src/main.c#L252) for instance.

joelguittet replied:
@rlubos my URL (with this example) is NOT constant.
At some point I have the following:
```
request.url = malloc(....)
strncpy(request.url, ...)
```

and later:
```
free(request.url);
```

The library is not touching this, but my application is.

joelguittet replied:
Actually to void the warnings at compilation time I need to do:
```
free((void *)request.url));
```

The compilation has no warning, but the static code analysis complains about discard of the const qualifier...

rlubos replied:
I see your point, but to me it should be up to the application to either:
* Do the casting as you suggest - it's the application which sets the pointer so it knows best that it's dynamically allocated (and thus not constant) and it's safe to remove `const` qualifier in such case,
* or have a separate `char *` pointer variable to remember the allocated memory address, if you concern about the static analysis.

`const` qualifier inside `struct http_request` fields is just a promise to the application that the library won't modify the memory pointed by that field. And because of that we can safely use string constants for instance, which seams to be heavily used across the sample code given CI results.

joelguittet replied:
The second option you mentioned is not bad :-) I will check for this.
I know the interest for the const qualifier, just need to properly find the right way to work. Else not only the `char*` should be const, but all the other fields probably too ;-)

joelguittet commented:
Solution provided by @rlubos working here, closing this PR and related issue so that we can keep the const qualifier as-is.
